# Recommendation System Final Assignment by Prof. Matsumura


## Author Information
Name: Junjie Xu

## Contents of Assignment

User-based filtering algorithms, like the 'getRecommendations()' function shown in Lesson 8, are inefficient because they compare the user with all other users every time a recommendation is needed. Follow the steps below to improve this.

## Step 1
Here is the code for step 1, calculateSimilarPersons()

(1.	Create a function 'calculateSimilarPersons()' that precomputes the similarity of users. The argument of this function is a dictionary of movies and their ratings, and the data structure is the same as the variable 'critics' shown in Lesson 8. The return value is a dictionary of similarities between all users, and the data structure is the same as the return value of 'calculateSimilarItems()' shown in Lesson 9.)


### calculateSimilarPersons()

In [1]:
# A dictionary of movie critics and their ratings of a small
# set of movies
critics={'Lisa Rose': {'Lady in the Water': 2.5, 'Snakes on a Plane': 3.5,
 'Just My Luck': 3.0, 'Superman Returns': 3.5, 'You, Me and Dupree': 2.5,
 'The Night Listener': 3.0},
'Gene Seymour': {'Lady in the Water': 3.0, 'Snakes on a Plane': 3.5,
 'Just My Luck': 1.5, 'Superman Returns': 5.0, 'The Night Listener': 3.0,
 'You, Me and Dupree': 3.5},
'Michael Phillips': {'Lady in the Water': 2.5, 'Snakes on a Plane': 3.0,
 'Superman Returns': 3.5, 'The Night Listener': 4.0},
'Claudia Puig': {'Snakes on a Plane': 3.5, 'Just My Luck': 3.0,
 'The Night Listener': 4.5, 'Superman Returns': 4.0,
 'You, Me and Dupree': 2.5},
'Mick LaSalle': {'Lady in the Water': 3.0, 'Snakes on a Plane': 4.0,
 'Just My Luck': 2.0, 'Superman Returns': 3.0, 'The Night Listener': 3.0,
 'You, Me and Dupree': 2.0},
'Jack Matthews': {'Lady in the Water': 3.0, 'Snakes on a Plane': 4.0,
 'The Night Listener': 3.0, 'Superman Returns': 5.0, 'You, Me and Dupree': 3.5},
'Toby': {'Snakes on a Plane':4.5,'You, Me and Dupree':1.0,'Superman Returns':4.0}}

from math import sqrt

# Returns a distance-based similarity score for person1 and person2
def sim_distance(prefs,person1,person2):
  # Get the list of shared_items
  si={}
  for item in prefs[person1]:
    if item in prefs[person2]:
       si[item]=1

  # if they have no ratings in common, return 0
  if len(si)==0: return 0

  # Add up the squares of all the differences
  sum_of_squares=sum([pow(prefs[person1][item]-prefs[person2][item],2)
                      for item in si])

  return 1/(1+sqrt(sum_of_squares))

# Returns the Pearson correlation coefficient for p1 and p2
def sim_pearson(prefs,p1,p2):
  # Get the list of mutually rated items
  si={}
  for item in prefs[p1]:
    if item in prefs[p2]: si[item]=1

  # Find the number of elements
  n=len(si)

  # if they have no ratings in common, return 0
  if n==0: return 0

  # Add up all the preferences
  sum1=sum([prefs[p1][it] for it in si])
  sum2=sum([prefs[p2][it] for it in si])

  # Sum up the squares
  sum1Sq=sum([pow(prefs[p1][it],2) for it in si])
  sum2Sq=sum([pow(prefs[p2][it],2) for it in si])

  # Sum up the products
  pSum=sum([prefs[p1][it]*prefs[p2][it] for it in si])

  # Calculate Pearson score
  num=pSum - (sum1*sum2/n)
  den=sqrt((sum1Sq - pow(sum1,2)/n)*(sum2Sq - pow(sum2,2)/n))
  if den==0: return 0
  r=num/den

  return r

  return 1/(1+sqrt(sum_of_squares)) 

def topSimilarityPerson(prefs, person, similarity = sim_pearson):
  # return a list of Pearson scores between person and other people in prefs in descending order of scores
  scores=[(similarity(prefs,person,other),other)
                  for other in prefs if other!=person]

  # Sort the list for all users so that the similarity appears in descending order of scores
  scores.sort()
  # reverse (user, score) to (score, user)
  scores.reverse()
  return scores
  
def calculateSimilarPersons(prefs):
  # precomputes the similarity of users
  result={}

  personPrefs=prefs
  c=0
  for person in personPrefs:
    # Status updates for large datasets
    c+=1
    if c%100==0: print("%d / %d" % (c,len(personPrefs)))
    # Find the most similar items to this one
    scores=topSimilarityPerson(personPrefs,person)
    result[person]=scores
  return result


## Step 2

Here is the code for step 2 

(2.	Assign the similarity data between users calculated above to the variable 'personsim' and check the contents of it.)

### Assign similarity data between users in 'personsim'

In [2]:
personsim=calculateSimilarPersons(critics)
personsim

{'Claudia Puig': [(1.0, 'Michael Phillips'),
  (0.8934051474415647, 'Toby'),
  (0.5669467095138411, 'Mick LaSalle'),
  (0.5669467095138396, 'Lisa Rose'),
  (0.31497039417435607, 'Gene Seymour'),
  (0.02857142857142857, 'Jack Matthews')],
 'Gene Seymour': [(0.963795681875635, 'Jack Matthews'),
  (0.41176470588235276, 'Mick LaSalle'),
  (0.39605901719066977, 'Lisa Rose'),
  (0.38124642583151164, 'Toby'),
  (0.31497039417435607, 'Claudia Puig'),
  (0.20459830184114206, 'Michael Phillips')],
 'Jack Matthews': [(0.963795681875635, 'Gene Seymour'),
  (0.7470178808339965, 'Lisa Rose'),
  (0.66284898035987, 'Toby'),
  (0.21128856368212925, 'Mick LaSalle'),
  (0.13483997249264842, 'Michael Phillips'),
  (0.02857142857142857, 'Claudia Puig')],
 'Lisa Rose': [(0.9912407071619299, 'Toby'),
  (0.7470178808339965, 'Jack Matthews'),
  (0.5940885257860044, 'Mick LaSalle'),
  (0.5669467095138396, 'Claudia Puig'),
  (0.40451991747794525, 'Michael Phillips'),
  (0.39605901719066977, 'Gene Seymour')],
 'M

{'Claudia Puig': [(1.0, 'Michael Phillips'),
  (0.8934051474415647, 'Toby'),
  (0.5669467095138411, 'Mick LaSalle'),
  (0.5669467095138396, 'Lisa Rose'),
  (0.31497039417435607, 'Gene Seymour'),
  (0.02857142857142857, 'Jack Matthews')],
 'Gene Seymour': [(0.963795681875635, 'Jack Matthews'),
  (0.41176470588235276, 'Mick LaSalle'),
  (0.39605901719066977, 'Lisa Rose'),
  (0.38124642583151164, 'Toby'),
  (0.31497039417435607, 'Claudia Puig'),
  (0.20459830184114206, 'Michael Phillips')],
 'Jack Matthews': [(0.963795681875635, 'Gene Seymour'),
  (0.7470178808339965, 'Lisa Rose'),
  (0.66284898035987, 'Toby'),
  (0.21128856368212925, 'Mick LaSalle'),
  (0.13483997249264842, 'Michael Phillips'),
  (0.02857142857142857, 'Claudia Puig')],
 'Lisa Rose': [(0.9912407071619299, 'Toby'),
  (0.7470178808339965, 'Jack Matthews'),
  (0.5940885257860044, 'Mick LaSalle'),
  (0.5669467095138396, 'Claudia Puig'),
  (0.40451991747794525, 'Michael Phillips'),
  (0.39605901719066977, 'Gene Seymour')],
 'Michael Phillips': [(1.0, 'Claudia Puig'),
  (0.40451991747794525, 'Lisa Rose'),
  (0.20459830184114206, 'Gene Seymour'),
  (0.13483997249264842, 'Jack Matthews'),
  (-0.2581988897471611, 'Mick LaSalle'),
  (-1.0, 'Toby')],
 'Mick LaSalle': [(0.9244734516419049, 'Toby'),
  (0.5940885257860044, 'Lisa Rose'),
  (0.5669467095138411, 'Claudia Puig'),
  (0.41176470588235276, 'Gene Seymour'),
  (0.21128856368212925, 'Jack Matthews'),
  (-0.2581988897471611, 'Michael Phillips')],
 'Toby': [(0.9912407071619299, 'Lisa Rose'),
  (0.9244734516419049, 'Mick LaSalle'),
  (0.8934051474415647, 'Claudia Puig'),
  (0.66284898035987, 'Jack Matthews'),
  (0.38124642583151164, 'Gene Seymour'),
  (-1.0, 'Michael Phillips')]}

## Step 3

Here is the code for step 3

(3.	Create the function 'getRecommendations2()' by modifying the function 'getRecommendations()' as follows.
o	Add the similarity data between users as an argument, and do not calculate the similarity between users in this function.
o	Only the five users with the highest similarity will be used to generate recommendations.)


### getRecommendations2()

In [3]:
# Gets recommendations for a person by using a weighted average
# of every other user's rankings
def getRecommendations2(prefs,personsim,person):
  totals = {}
  simSums = {}
  topPersonsim = dict(zip([user[1] for user in personsim[person]], [user[0] for user in personsim[person]]))


  for other in prefs:
    for item in prefs[other]:

      # only score movies I haven't seen yet
      if item not in prefs[person] or prefs[person][item]==0:
        # Similarity * Score
        totals.setdefault(item,0)
        sim = topPersonsim[other]
        # ignore scores of zero or lower
        if sim <= 0 : 
          sim = 0
        totals[item]+=prefs[other][item]*sim
        # Sum of similarities
        simSums.setdefault(item,0)
        simSums[item]+=sim

  # Create the normalized list
  rankings=[((total/simSums[item],item) if simSums[item] != 0 else (0,item)) for item,total in totals.items()]

  # Return the sorted list
  rankings.sort()
  rankings.reverse()
  return rankings

In [4]:
print(getRecommendations2(critics, personsim, 'Toby'))

[(3.3477895267131017, 'The Night Listener'), (2.8325499182641614, 'Lady in the Water'), (2.530980703765565, 'Just My Luck')]


[(3.3477895267131017, 'The Night Listener'), (2.8325499182641614, 'Lady in the Water'), (2.530980703765565, 'Just My Luck')]

## Step 4
Here is the code for step 4
(4.	Load the MovieLens dataset used in Lesson 9, assign it to the variable 'critics', and check the behavior of 'calculateSimilarPersons()' and 'getRecommendations2()'.)

### Download and load dataset directly from MovieLens's website to colab
Note that in my step 4, this code loads MovieLens dataset in colab without mount the drive

In [5]:
# download MovieLens dataset
!wget "http://files.grouplens.org/datasets/movielens/ml-latest-small.zip" -O ml-latest-small.zip
!unzip "ml-latest-small.zip"


# load MovieLens dataset
import csv
def loadMovieLens(path='/content/ml-latest-small'):

  # Get movie titles
  movies={}
  f1=open(path+'/movies.csv')
  reader1=csv.reader(f1)
  header1=next(reader1)
  for row in reader1:
    (movieId,title,_)=row
    movies[movieId]=title
  f1.close()

  # Load data
  f2=open(path+'/ratings.csv')
  reader2=csv.reader(f2)
  header2=next(reader2)
  prefs={}
  for row in reader2:
    (userId,movieId,rating,timestamp)=row
    prefs.setdefault(userId,{})
    prefs[userId][movies[movieId]]=float(rating)
  return prefs




--2020-12-31 05:46:35--  http://files.grouplens.org/datasets/movielens/ml-latest-small.zip
Resolving files.grouplens.org (files.grouplens.org)... 128.101.65.152
Connecting to files.grouplens.org (files.grouplens.org)|128.101.65.152|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 978202 (955K) [application/zip]
Saving to: ‘ml-latest-small.zip’

ml-latest-small.zip 100%[===================>] 955.28K  --.-KB/s    in 0.1s    

2020-12-31 05:46:35 (7.64 MB/s) - ‘ml-latest-small.zip’ saved [978202/978202]

Archive:  ml-latest-small.zip
   creating: ml-latest-small/
  inflating: ml-latest-small/links.csv  
  inflating: ml-latest-small/tags.csv  
  inflating: ml-latest-small/ratings.csv  
  inflating: ml-latest-small/README.txt  
  inflating: ml-latest-small/movies.csv  


--2020-12-31 05:46:35--  http://files.grouplens.org/datasets/movielens/ml-latest-small.zip
Resolving files.grouplens.org (files.grouplens.org)... 128.101.65.152
Connecting to files.grouplens.org (files.grouplens.org)|128.101.65.152|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 978202 (955K) [application/zip]
Saving to: ‘ml-latest-small.zip’

ml-latest-small.zip 100%[===================>] 955.28K  --.-KB/s    in 0.1s    

2020-12-31 05:46:35 (7.64 MB/s) - ‘ml-latest-small.zip’ saved [978202/978202]

Archive:  ml-latest-small.zip
   creating: ml-latest-small/
  inflating: ml-latest-small/links.csv  
  inflating: ml-latest-small/tags.csv  
  inflating: ml-latest-small/ratings.csv  
  inflating: ml-latest-small/README.txt  
  inflating: ml-latest-small/movies.csv 

In [6]:
prefs=loadMovieLens()

### test calculateSimilarPersons()

In [ ]:
moviePersonsim=calculateSimilarPersons(prefs)
moviePersonsim

### test getRecommendations2()

In [11]:
getRecommendations2(prefs, moviePersonsim, '87')[0:30]

[(5.0, 'Zeitgeist: Moving Forward (2011)'),
 (5.0, 'Wow! A Talking Fish! (1983)'),
 (5.0, 'Wonder Woman (2009)'),
 (5.0, 'Woman Under the Influence, A (1974)'),
 (5.0, 'Woman Is a Woman, A (femme est une femme, Une) (1961)'),
 (5.0, 'Winter in Prostokvashino (1984)'),
 (5.0, 'Winnie the Pooh and the Day of Concern (1972)'),
 (5.0, 'Winnie the Pooh Goes Visiting (1971)'),
 (5.0, 'Winnie Pooh (1969)'),
 (5.0, 'Wings, Legs and Tails (1986)'),
 (5.0, 'When Worlds Collide (1951)'),
 (5.0, 'What Men Talk About (2010)'),
 (5.0, 'Watermark (2014)'),
 (5.0, 'War for the Planet of the Apes (2017)'),
 (5.0, 'Vovka in the Kingdom of Far Far Away (1965)'),
 (5.0, 'Very Potter Sequel, A (2010)'),
 (5.0, 'Vampire in Venice (Nosferatu a Venezia) (Nosferatu in Venice) (1986)'),
 (5.0, 'Vagabond (Sans toit ni loi) (1985)'),
 (5.0, 'Vacations in Prostokvashino (1980)'),
 (5.0, 'Unfaithfully Yours (1948)'),
 (5.0, "Tyler Perry's I Can Do Bad All by Myself (2009)"),
 (5.0, 'Two Family House (2000)'),
 (5.0

[(5.0, 'Zeitgeist: Moving Forward (2011)'),
 (5.0, 'Wow! A Talking Fish! (1983)'),
 (5.0, 'Wonder Woman (2009)'),
 (5.0, 'Woman Under the Influence, A (1974)'),
 (5.0, 'Woman Is a Woman, A (femme est une femme, Une) (1961)'),
 (5.0, 'Winter in Prostokvashino (1984)'),
 (5.0, 'Winnie the Pooh and the Day of Concern (1972)'),
 (5.0, 'Winnie the Pooh Goes Visiting (1971)'),
 (5.0, 'Winnie Pooh (1969)'),
 (5.0, 'Wings, Legs and Tails (1986)'),
 (5.0, 'When Worlds Collide (1951)'),
 (5.0, 'What Men Talk About (2010)'),
 (5.0, 'Watermark (2014)'),
 (5.0, 'War for the Planet of the Apes (2017)'),
 (5.0, 'Vovka in the Kingdom of Far Far Away (1965)'),
 (5.0, 'Very Potter Sequel, A (2010)'),
 (5.0, 'Vampire in Venice (Nosferatu a Venezia) (Nosferatu in Venice) (1986)'),
 (5.0, 'Vagabond (Sans toit ni loi) (1985)'),
 (5.0, 'Vacations in Prostokvashino (1980)'),
 (5.0, 'Unfaithfully Yours (1948)'),
 (5.0, "Tyler Perry's I Can Do Bad All by Myself (2009)"),
 (5.0, 'Two Family House (2000)'),
 (5.0, 'True Stories (1986)'),
 (5.0, 'Troll 2 (1990)'),
 (5.0, 'Travels of an Ant (1983)'),
 (5.0, 'Trailer Park Boys (1999)'),
 (5.0, 'Tokyo Tribe (2014)'),
 (5.0, 'Three from Prostokvashino (1978)'),
 (5.0, 'Thousand Clowns, A (1965)'),
 (5.0, 'There Once Was a Dog (1982)')]